Lib/inspect.py

```diff
@@ -2609,28 +2619,27 @@ def _signature_from_callable(obj, *,
 
         # First, let's see if it has an overloaded __call__ defined
         # in its metaclass
-        call = _signature_get_user_defined_method(type(obj), '__call__')
-        if call is not None:
-            sig = _get_signature_of(call)
+        call = getattr_static(type(obj), '__call__', None)
+        if call is not None and not isinstance(call, _NonUserDefinedCallables):
+            call = _descriptor_get(call, obj)
+            return _get_signature_of(call)
         else:
-            factory_method = None
             new = _signature_get_user_defined_method(obj, '__new__')
```

Comment:
It seems like `__new__` is the only exemption that continues to use this original method. All the others are replaced with the new logic which uses `getattr_static` to avoid descriptor resolving. Is there a reason for that? Can we simply change `_signature_get_user_defined_method`?

Reply:
I think `_signature_get_user_defined_method` is still a pretty self-explanatory name for the method. The only usage for this private method is to get user defined `__new__`, `__init__` and `__call__` methods. Can we do the check logic inside the method, asserting the input method to check, and use `getattr()` / `getattr_static()` accordingly? The only difference between `__new__` and the other two is which function to use right? We don't even need the `_descriptor_get` function, and we don't need to change the logic in this function.

Reply:
We can do this, but it will split the logic on two parts, 600 lines apart one from other. There are other differences:

* Note that the code for `__new__` is followed by `break` (to call `_signature_bound_method()` if `skip_bound_arg` is true), while the code for other methods immediately returns. So some logic will be here, and some in `_signature_get_user_defined_method()`.

* The code for non-type's `__call__` does not have restrictions for builtin types. In future, I want to loosen restrictions for other methods too, so `isinstance(meth, _NonUserDefinedCallables)` will be different for every method, and we may need to check the type of the descriptor before resolving it.

Reply:
Okay I still think the logic is hard to follow. It took me a while to figure out that the key thing changed is that `init` and `call` are currently bound methods, so they do not need the strip of the first parameter - this logic for `new` however, is far down.

I still believe that we should put the new descriptor/getattr_static logic into `_signature_bound_method()` because that's exactly what they do. You still have the old function and that's having two super similar routine far away from each other.

I think the logic in `_signature_from_callable` should be as simple as

```python
call = _signature_get_user_defined_method(type(obj), '__call__')
new = _signature_get_user_defined_method(obj, '__new__')
init = _signature_get_user_defined_method(obj, '__init__')
```

However, we need to deal with the bound-issue, which is what `slot_tp_call` does - we can just return another indicator:

```python
call, unbound = _signature_get_user_defined_method(type(obj), '__call__')
```

We would significantly simplify the logic in `_signature_from_callable` and get what we need (function and whether it's bounded) with a single function call.

Then, inside `isinstance(obj, type)` check, before `if sig is None:`, we can simply do

```python
if sig is not None:
    if unbound and skip_bound_arg:
        return _signature_bound_method(sig)
    return sig
```

Then of course the following code does not need the if check protection as we ensured `sig is not None` at that point.

This way, we have all the related logic in two places:

1. `_signature_get_user_defined_method` that "magically" does the check and resolves the descriptor.
2. A continuous block in `if isinstance(obj, type):` that solves everything.

Reply:
My goal was to minimize the diffs for easier review and do refactoring in the following PR. But if you prefer, I do some refactoring in this PR (complete refactoring will be possible after making some other behavior changes).

I tried to implement your suggestions, except returning `unbound`, because I do not see how it can help.